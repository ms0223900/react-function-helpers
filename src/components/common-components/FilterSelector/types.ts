import { Callback } from "all-common-types";
import { ChangeEvent } from "react";

export interface SingleSelectorOption {
  text: string
  value: string
  [x: string]: any
}
export type SelectorOptions = SingleSelectorOption[]
export type GetSelectedOptionFn = (selected: SingleSelectorOption) => any

export interface SingleSelectProps {
  option: SingleSelectorOption
  isSelected: boolean
  clickFn?: Callback
}

export interface SelectsInFilterProps {
  filterFn?: (e: ChangeEvent<HTMLInputElement>) => any 
  selectFn?: (index: number) => any
  closeSelectsFn?: Callback
  filterInput: string
  filteredOptions: SelectorOptions
  selectedIndex: number | undefined
}

export interface FilterSelectorProps extends SelectsInFilterProps {
  isDisplaySelects: boolean
  selectedText: string
  options: SelectorOptions
  toggleDisplaySelectsFn?: Callback
}

export interface FilterSelectorContainerProps {
  defaultSelectedText?: string
  options: SelectorOptions
  getSelectedOptionFn?: GetSelectedOptionFn
}

export interface FilterSelectorContainerStates {
  selectedIndex: number | undefined
  selectedText: string
  isDisplaySelects: boolean
  filterInput: string
  filteredOptions: SelectorOptions
}