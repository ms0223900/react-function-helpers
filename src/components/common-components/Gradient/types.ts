import { ID, Callback } from "all-common-types";

export type ColorPercent = 'color' | 'percent'

export interface GradientItemValue {
  id: ID
  color: string
  percent: number | string
}

export interface ColorPinProps {
  position: {
    x: number,
    y: number,
  },
  gradientItemValue: GradientItemValue,
  isSelected: boolean,
  clickFn?: Callback
}

export interface ColorPinListProps {
  gradientItemValueList: GradientItemValue[]
  setSelectedPercent?: (id: ID, percent: number) => any
}
