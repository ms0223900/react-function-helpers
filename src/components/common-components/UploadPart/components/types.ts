import { ChangeEvent, FormEvent, ReactNode, ReactNodeArray } from "react";
import { Callback, ImagePreviewFile } from "../lib/types";
import { DragDropFileBoxProps } from "./DragDropFileBox";

export interface UploadFormProps extends DragDropFileBoxProps {
  isUploadAvailble?: boolean;
  children?: ReactNode | ReactNodeArray;
  onSubmit: Callback;
  onChangeFiles: (e: ChangeEvent<HTMLInputElement>) => any;
}

export interface PreviewImageItemProps {
  index: number
  file: ImagePreviewFile
  previewImageProps?: React.ImgHTMLAttributes<HTMLImageElement>
  onRemoveImage?: (index: number) => Callback
}

export interface PreviewPartProps extends Pick<PreviewImageItemProps, 'previewImageProps' | 'onRemoveImage'> {
  previewImageList: ImagePreviewFile[];
}
