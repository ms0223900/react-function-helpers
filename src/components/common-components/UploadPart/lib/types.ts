export type Callback = (args?: any) => any;

export interface ImagePreviewFile {
  name: string;
  src: string;
}

// vscode 不會報錯，此行是給codesandbox用的
export type RequestInit = any;
