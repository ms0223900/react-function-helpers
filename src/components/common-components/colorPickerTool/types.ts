import { Theme } from "@material-ui/core";

export type PaletteKeys = keyof Theme['palette']
export interface ThemeCustomizableColors {
  primary?: string, 
  secondary?: string,
}
