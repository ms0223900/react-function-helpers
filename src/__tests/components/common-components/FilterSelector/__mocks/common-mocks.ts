import { SelectPayload, ResetSelectActionPayload, FitlerActionPayload, SetStateActionPayload } from "../../../../../components/common-components/FilterSelector/actions";
import { SingleSelectorOption, FilterSelectorContainerStates, FilterSelectorContainerProps } from "../../../../../components/common-components/FilterSelector/types";

export const singleSelectorOption: SingleSelectorOption = {
  value: 'a',
  text: 'A',
};

export const selectPayload: SelectPayload = {
  selectedIndex: 0,
  options: [],
};

export const resetSelectActionPayload: ResetSelectActionPayload = [singleSelectorOption];

export const fitlerActionPayload: FitlerActionPayload = {
  filterInput: '',
  options: [],
};

export const setStateActionPayload: SetStateActionPayload = {
  selectedText: 'cjk',
};

export const customInitState: FilterSelectorContainerStates = {
  selectedIndex: 10,
  selectedText: 'abc',
  isDisplaySelects: true,
  filterInput: 'aaa',
  filteredOptions: [
    singleSelectorOption,
    singleSelectorOption,
  ]
};

export const filterSelectorContainerProps: FilterSelectorContainerProps = {
  options: [singleSelectorOption],
  defaultSelectedText: 'bb',
};