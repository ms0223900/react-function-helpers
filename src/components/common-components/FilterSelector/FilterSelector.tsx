import { 
  Box, 
  Button, 
  makeStyles,
  Typography,
} from '@material-ui/core';
import React from 'react';
import Selects from './SelectsInFilter';
import { FilterSelectorProps } from './types';

const useStyles = makeStyles(theme => ({
  root: {
    marginTop: theme.spacing(1),
    marginBottom: theme.spacing(1),
    position: 'relative',
  },
  selectButton: {
    width: '100%',
  },
}));

const FilterSelector = (props: FilterSelectorProps) => {
  const {
    selectedText, 
    isDisplaySelects,
    toggleDisplaySelectsFn,
  } = props;
  const classes = useStyles();

  return (
    <Box className={ classes.root } >
      <Button 
        className={ classes.selectButton }
        variant={ 'outlined' } 
        onClick={toggleDisplaySelectsFn}
      >
        { selectedText }
        <Typography variant={ 'body1' }>
          { ' â–¾ ' }
        </Typography>
      </Button>
      {isDisplaySelects && (
        <Selects 
          {...props}
          closeSelectsFn={toggleDisplaySelectsFn}
        />
      )}
    </Box>
  );
};


export default FilterSelector;
