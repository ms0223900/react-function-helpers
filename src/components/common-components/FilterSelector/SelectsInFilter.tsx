import React, { memo, useCallback, useRef, createRef, useEffect, forwardRef, Ref, MutableRefObject } from 'react';
import { 
  TextField, 
  List,
  ListItem,
  ListItemText,
  Paper,
  CircularProgress,
  makeStyles,
  Box,
} from '@material-ui/core';
import { SingleSelectProps, SelectsInFilterProps } from './types';
import { useStylesSelectsInFilter } from './styles';
import useScrollToSingleSelect from './useScrollToSingleSelect';

export const NoOptionsInfo = ({ input }: {
  input: string
}) => {
  return (
    <ListItem>
      <ListItemText primary={ `no result matched ${ input }` } />
    </ListItem>);
};

export const SingleSelect = forwardRef(({
  option, isSelected, clickFn
}: SingleSelectProps, ref: Ref<HTMLDivElement>) => {
  return (
    <ListItem  
      ref={ref}
      selected={ isSelected }
      onClick={ clickFn } 
      button
    >
      <ListItemText primary={ option.text } />
    </ListItem>
  );
});

export const Selects = ({ 
  filterInput, 
  filteredOptions, 
  selectedIndex, 
  filterFn, 
  selectFn,
  closeSelectsFn 
}: SelectsInFilterProps) => {
  const listRefs = React.useRef(filteredOptions.map((op) => (
    React.createRef<HTMLDivElement>()
  )));
  const classes = useStylesSelectsInFilter();

  const handleSelectRoute = useCallback((index: number) => {
    return () => {
      selectFn && selectFn(index);
    };
  }, [selectFn]);

  const handleClose = useCallback(() => {
    closeSelectsFn && closeSelectsFn();
  }, [closeSelectsFn]);

  useScrollToSingleSelect(listRefs, selectedIndex);

  return (
    <Box className={ classes && classes.root }>
      <Paper>
        <Box padding={1}>
          <TextField 
            fullWidth 
            autoFocus={ true }
            variant={'outlined'}
            onChange={ filterFn }
            value={ filterInput } />
        </Box>
        <List className={classes.selects}>
          { filteredOptions.length > 0 ? (
            filteredOptions.map((option, i) => (
              <SingleSelect 
                key={i}
                ref={listRefs.current[i]}
                option={option}
                isSelected={selectedIndex === i}
                clickFn={handleSelectRoute(i)}  />
            ))
          ) : (
            <NoOptionsInfo input={filterInput} />
          ) }
        </List>
      </Paper>
      <Box 
        id={'selectsInFilterBG'}
        className={classes.clickBG} 
        onClick={handleClose}
      />
    </Box>
  );
};

export default memo(Selects);