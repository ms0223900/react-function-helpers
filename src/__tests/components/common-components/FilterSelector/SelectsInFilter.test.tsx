import React from 'react';
import { shallow } from 'enzyme';
import renderer, { act } from 'react-test-renderer';
import SelectsInFilter, { SingleSelect, NoOptionsInfo } from '../../../../components/common-components/FilterSelector/SelectsInFilter';
import { selectsInFilterProps, singleSelectorOption, singleSelectProps } from './__mocks/common-mocks';
import { TextField } from '@material-ui/core';
import { SelectorOptions } from '../../../../components/common-components/FilterSelector/types';

describe('test components in SelectsInFilter', () => {
  it('should render correctly', () => {
    const tree = renderer.create(
      <SelectsInFilter {...selectsInFilterProps} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('NoOptionsInfo should render correctly', () => {
    const tree = renderer.create(
      <NoOptionsInfo input={''} />
    ).toJSON();
    expect(tree).toMatchSnapshot(); 
  });

  it('SingleSelect should render correctly', () => {
    const tree = renderer.create(
      <SingleSelect {...singleSelectProps} />
    ).toJSON();
    expect(tree).toMatchSnapshot(); 
  });
});

describe('test SelectsInFilter components determined by props.filteredOptions.length', () => {
  it('props.filteredOptions.length > 0 condition(render SingleSelect)', () => {
    const filteredOptions = [singleSelectorOption];
    const wrapper = shallow(
      <SelectsInFilter 
        {...selectsInFilterProps}
        filteredOptions={filteredOptions} />
    );
    expect(wrapper.find(SingleSelect)).toHaveLength(filteredOptions.length);
  });

  it('props.filteredOptions.length > 0 condition(render SingleSelect)', () => {
    const filteredOptions: SelectorOptions = [];
    const wrapper = shallow(
      <SelectsInFilter 
        {...selectsInFilterProps}
        filteredOptions={filteredOptions} />
    );
    expect(wrapper.find(SingleSelect)).toHaveLength(0);
    expect(wrapper.find(NoOptionsInfo)).toHaveLength(1);
  });
});

describe('test functions should be trigger correctly', () => {
  it('test handleClose', () => {
    const closeSelectsFn = jest.fn();
    const wrapper = shallow(
      <SelectsInFilter 
        {...selectsInFilterProps}
        closeSelectsFn={closeSelectsFn} />
    );

    expect(closeSelectsFn).not.toBeCalled();
    act(() => {
      const { onClick } = wrapper.find('#selectsInFilterBG').props();
      onClick && onClick({} as any);
    });
    
    expect(closeSelectsFn).toBeCalled();
  });

  it('test handleSelectRoute', () => {
    const index = 1;
    const selectFn = jest.fn();
    const wrapper = shallow(
      <SelectsInFilter 
        {...selectsInFilterProps}
        selectFn={selectFn}
        filteredOptions={[
          singleSelectorOption,
          singleSelectorOption,
        ]} />
    );

    expect(selectFn).not.toBeCalled();
    act(() => {
      const { clickFn } = wrapper.find(SingleSelect).at(index).props();
      clickFn && clickFn(index);
    });

    expect(selectFn).toBeCalledWith(index);
  });

  it('test filterFn', () => {
    const e = { target: { value: '', }, } as React.ChangeEvent<HTMLInputElement>;
    const filterFn = jest.fn();
    const wrapper = shallow(
      <SelectsInFilter 
        {...selectsInFilterProps}
        filterFn={filterFn}
        filteredOptions={[
          singleSelectorOption,
          singleSelectorOption,
        ]} />
    );

    expect(filterFn).not.toBeCalled();
    act(() => {
      const { onChange } = wrapper.find(TextField).props();
      onChange && onChange(e);
    });
    expect(filterFn).toBeCalled();
  });

});
