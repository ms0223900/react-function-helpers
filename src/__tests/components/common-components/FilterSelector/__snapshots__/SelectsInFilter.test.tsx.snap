// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test components in SelectsInFilter NoOptionsInfo should render correctly 1`] = `
<li
  className="MuiListItem-root MuiListItem-gutters"
  disabled={false}
>
  <div
    className="MuiListItemText-root"
  >
    <span
      className="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
    >
      no result matched 
    </span>
  </div>
</li>
`;

exports[`test components in SelectsInFilter SingleSelect should render correctly 1`] = `
<div
  aria-disabled={false}
  className="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
  onBlur={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  role="button"
  tabIndex={0}
>
  <div
    className="MuiListItemText-root"
  >
    <span
      className="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
    >
      A
    </span>
  </div>
</div>
`;

exports[`test components in SelectsInFilter should render correctly 1`] = `
<div
  className="MuiBox-root MuiBox-root-5 makeStyles-root-1"
>
  <div
    className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="MuiBox-root MuiBox-root-34"
    >
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={true}
            className="MuiInputBase-input MuiOutlinedInput-input"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-75 MuiOutlinedInput-notchedOutline"
            style={
              Object {
                "paddingLeft": 8,
              }
            }
          >
            <legend
              className="PrivateNotchedOutline-legend-76"
              style={
                Object {
                  "width": 0.01,
                }
              }
            >
              <span
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "&#8203;",
                  }
                }
              />
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <ul
      className="MuiList-root makeStyles-selects-2 MuiList-padding"
    >
      <li
        className="MuiListItem-root MuiListItem-gutters"
        disabled={false}
      >
        <div
          className="MuiListItemText-root"
        >
          <span
            className="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
          >
            no result matched 
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div
    className="MuiBox-root MuiBox-root-128 makeStyles-clickBG-3"
    id="selectsInFilterBG"
    onClick={[Function]}
  />
</div>
`;
